spec/models/user_spec.rb

```diff
@@ -1,5 +1,73 @@
-require 'rails_helper'
+require "rails_helper"
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe "enum for role" do
+    it {is_expected.to define_enum_for(:role).with_values(admin: 0, customer: 1)}
+  end
+
+  describe "associations" do
+    it {is_expected.to have_many(:orders).dependent(:destroy)}
+    it {is_expected.to have_many(:comments).dependent(:destroy)}
+  end
+
+  describe "validates" do
+    context "validate presence" do
+      it {is_expected.to validate_presence_of :email}
+      it {is_expected.to validate_presence_of :name}
+      it {is_expected.to validate_presence_of :password}
+    end
+
+    context "validate length of name" do
+      it {is_expected.to validate_length_of(:name).is_at_least(Settings.user.name_validates.name_min_length).is_at_most(Settings.user.name_validates.name_max_length)}
+    end
+
+    context "validate length of password" do
+      it {is_expected.to validate_length_of(:password).is_at_least(Settings.user.password_validates.password_min_length)}
+    end
+
+    context "validate length of address" do
+      it {is_expected.to validate_length_of(:address).is_at_least(Settings.user.address_min_length).is_at_most(Settings.user.address_max_length)}
+    end
+
+    context "validate length of phone" do
+      it {is_expected.to validate_length_of(:phone_num).is_at_least(Settings.user.min_phone).is_at_most(Settings.user.max_phone)}
+    end
+  end
+
+  describe "scope" do
+    let!(:user1) {FactoryBot.create :user, name: "Tien", role: 0}
+    let!(:user2) {FactoryBot.create :user, name: "Dien", phone_num: "0987365423"}
+    let!(:user3) {FactoryBot.create :user, name: "Nguyen"}
+
+    it "check scope user asc name" do
+      expect(User.asc_name).to eq([user2, user3, user1])
+    end
+
+    it "check scope user get all" do
+      expect(User.get_all).to eq([user2, user3])
+    end
+
+    it "check scope search by name" do
+      expect(User.search_by_name("Ti")).to eq([user1])
+    end
+
+    it "check scope search wrong name" do
+      expect(User.search_by_name("Tisdaasdasd")).to eq([])
+    end
+
+    it "check scope search by phone_num" do
+      expect(User.search_by_name("098")).to eq([user2])
+    end
+
+    it "check scope search wrong phone_num" do
+      expect(User.search_by_name("09812")).to eq([])
+    end
+  end
+
+  describe "downcase email" do
```

Comment:
xem link này xem sao https://stackoverflow.com/questions/16677718/how-to-test-models-callback-method-independently